src/main/java/handong/whynot/service/S3Service.java

```diff
@@ -0,0 +1,95 @@
+package handong.whynot.service;
+
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.amazonaws.services.s3.model.CannedAccessControlList;
+import com.amazonaws.services.s3.model.PutObjectRequest;
+import handong.whynot.dto.cloud.S3ResponseCode;
+import handong.whynot.exception.cloud.S3AclException;
+import handong.whynot.exception.cloud.S3InvalidFileTypeException;
+import handong.whynot.util.ImageType;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Optional;
+
+@Service
+@RequiredArgsConstructor
+public class S3Service {
+
+    @Value("${cloud.aws.s3.bucket}")
+    public String bucket;
+
+    private final AmazonS3Client amazonS3Client;
+
+    // S3 파일 업로드 수행
+    public String upload(MultipartFile multipartFile, String dirName) throws IOException {
+        File uploadFile = convert(multipartFile)
+                .orElseThrow(() -> new S3InvalidFileTypeException(S3ResponseCode.AWS_S3_UPLOAD_FAIL));
+
+        return upload(uploadFile, dirName);
+    }
+
+    // S3로 파일 업로드
+    private String upload(File uploadFile, String dirName) {
+        // S3에 저장된 파일 이름
+        final String fileName = dirName + "/" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm-ssSS")) + "-" + uploadFile.getName();
```

Comment:
fileName을 반환하는 부분은 따로 메서드로 빼거나 `BiFunction`을 이용해서 최상단에 선언해줘도 좋을 것 같습니다.
```suggestion
private static final BiFunction<String, File, String> FILE_NAME =
            (dirName, uploadFile) -> dirName + '/' + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm-ssSS")) + '-' + uploadFile.getName();

final String fileName = FILE_NAME.apply(dirName, uploadFile);
```

Reply:
BiFunction 을 이럴때 쓰는 거군요!! 반영하겠습니닷!